### PR TITLE
Feat update php parser

### PR DESCRIPTION
**Pull request by evilebottnawi**
Prepare to new release. Feel free to feedback/fix/commit in this branch

### CODE REVIEW

**Comment by evilebottnawi** [comment]
@azz can you help us with this? Why comments inside `class`?

**Comment by mgrip** [comment]
@evilebottnawi can you create an issue for this worth some more detail on what we're trying to accomplish?

**Comment by evilebottnawi** [comment]
@mgrip it is just PoC, attach comments from `comments`, which in future version will be in `program.comments` (all good parsers do this)

**Comment by evilebottnawi** [comment]
@mgrip you can see what all comments now printed, some on incorrect place

**Comment by mgrip** [comment]
I still think it would be good to have an issue open where we can discuss this more - I'm unclear what benefits we'd get from attaching comments from `program.comments`

**Comment by mgrip** [comment]
also are there other benefits we'd get from updating the parser?

**Comment by evilebottnawi** [comment]
@mgrip Examples:
```php
$var /* comment */ = /* comment */ 'string';
```
Right now is is impossible. Also add newlines between statement which have leading comments. Also in future version we don't have `doc` (i.e. comment) node in ast.

**Comment by evilebottnawi** [comment]
@mgrip Fixes with `public`, `exit` and `die`, `shortForm` echo - `<?= ?>` and etc

**Comment by evilebottnawi** [comment]
@mgrip i update PR, you can see what we don't have `doc` node anymore (in CI or run locally)

File: src/index.js [hunk 1]
```diff
@@ -48,7 +48,36 @@ const parsers = {
...
   php: {
     print,
-    massageAstNode: clean
+    massageAstNode: clean,
+    getCommentChildNodes(node) {
+      let children = node.children || node.body || node.traits;
+      if (!children) {
+        return [];
+      }
+      if (children.kind && children.kind === "block") {
```
Comment (nicoder) [inline src/index.js]: I guess this would be equivalent:`if (children.kind === "block") {`

do you prefer the longer version?

Comment (evilebottnawi) [reply src/index.js]: @nicoder It is PoC, but you are right :+1:

**Comment by mgrip** [comment]
@evilebottnawi would this fix https://github.com/prettier/plugin-php/issues/191 and/or https://github.com/prettier/plugin-php/issues/190?

**Comment by evilebottnawi** [comment]
@mgrip Yes and Yes

**Comment by evilebottnawi** [comment]
@mgrip yes, we can fix and implement new features using latest master version from git (it is allow early bug reporting and easy migrate on 3 version). Help welcome :+1: Need solve problem with comments some comments on incorrect places, need investigate why and if it is bug we will report about this.

**Comment by mgrip** [comment]
I am so confused haha. I have the branch checked out and am trying to debug through how this works but I feel like I'm missing something. How do the comments actually get printed now? For example the class node - it seems like there's a new `class.comments` and `class.leadingComments` attribute, how does `printComment()` know how to print those?

**Comment by mgrip** [comment]
Believe this is related https://github.com/glayzzle/php-parser/issues/135

**Comment by evilebottnawi** [comment]
@mgrip using offset in position :smile: Better debug logic and look how it is work :smile:

**Comment by andreiglingeanu** [comment]
Here's a quick bug that I discovered (in this branch):

Before:

```php
<?php

/**
 * test comment
 */

class Test
{
	public function testme($param)
	{
	    return null;
	}

}
```

After:

```php
/**
 * test comment
 */

<?php
class Test
{
	public function testme($param)
	{
		return null;
	}
}
```

When we start the file with a comment, it moves the `<?php` tag right before the `class` declaration, leaving the comment at the top. I believe this is a breaking change.

**Comment by mgrip** [comment]
thanks for the bug report @andreiglingeanu!

@evilebottnawi I'm making some good progress here, will push an update to the `feat-update-php-parser` branch at some point today. One thing I'm worried about though is that a lot of the built-in logic for comments seems specific to the estree ast https://github.com/prettier/prettier/blob/master/src/main/comments.js#L794 - not sure how big of a problem it will end up being, but what do you think?

**Comment by evilebottnawi** [comment]
@mgrip let's push here all and create issue in `prettier` around problem, seems something should be redefined inside `parser` option

**Comment by mgrip** [comment]
@evilebottnawi just updated - still one failing test, but let me know what you think so far!

File: src/printer.js [hunk 2]
```diff
@@ -3,6 +3,8 @@
 const docBuilders = require("prettier").doc.builders;
 const docUtils = require("prettier").doc.utils;
 const sharedUtil = require("prettier").util;
+//@TODO: super hacky, remove once comment functions are exported for plugin use
+const comments = require("../node_modules/prettier/src/main/comments");
```
Comment (mgrip) [inline src/printer.js]: https://github.com/prettier/prettier/issues/4177

**Comment by evilebottnawi** [comment]
@mgrip great work :+1: I'll see what I can fix

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 3]
```diff
@@ -106,17 +106,25 @@ class Foo
...
- * Some long comment *
- *********************/
-// This is a one-line c++ style comment
-/* This is a multi line comment
-   yet another line of comment */
-/* This is a multiple-lines comment block
-that spans over multiple
-lines */
-// This is a one-line shell-style comment
+    /**
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: I think we should't change comment output, print as is

Comment (evilebottnawi) [reply tests/comments/__snapshots__/jsfmt.spec.js.snap]: It is very bad practice

Comment (evilebottnawi) [reply tests/comments/__snapshots__/jsfmt.spec.js.snap]: you can tests how prettier handle comment using `playground`

Comment (mgrip) [reply tests/comments/__snapshots__/jsfmt.spec.js.snap]: 👍 I agree. The printComments function needs some more work, was just trying to get something to cover the most common cases for now but agreed its currently too opinionated

**Comment by evilebottnawi** [comment]
/cc @mgrip @czosel 
Tests passed (except `AST_COMPARE=1`). Roadmap (asap - we should solve this problem first before merge anything in master):
1. Normalize only `isDoc` comment, comments what are not `isDoc` print as is (example look prettier logic).
2. Some comments have incorrect places, need investigate and solve.
3. Update `cleaner`.
4. Add more tests around comments.

**Comment by evilebottnawi** [comment]
We can't continue work on this until `prettier` don't solve problem around api :disappointed:

**Comment by mgrip** [comment]
As far as step 2. goes, I did some debugging and I believe its because some of the logic to attach comment nodes in the right place is ast-specific (https://github.com/prettier/prettier/issues/4177#issuecomment-374392660). `decorateComment()` is working as expected (precending, following, and enclosing nodes get populated correctly), but then we need to have access to the "attach" part of the process to handle different node types and situations for comments

**Comment by mgrip** [comment]
I think this PR is pretty important for us to get over the line to get to a v1 - do we think we should focus on contributing to prettier and php-parser to get these dependencies solved before solving more issues here?

**Comment by evilebottnawi** [comment]
@mgrip yes, it will be great, but some issue can take many times before will be merged and release :disappointed:

**Comment by czosel** [comment]
I'd suggest we try to fix the things that block us in our dependencies. Doesn't really matter when something is released, we can just put the right commit in package.json (like we do right now with our prettier dependency).

**Comment by czosel** [comment]
@mgrip Awesome! Just in case you didn't find the right file in prettier yet - this is the util which is available inside plugins: https://github.com/prettier/prettier/blob/master/src/common/util-shared.js

**Comment by mgrip** [comment]
Just pushed a patch based on https://github.com/prettier/prettier/pull/4182 - let me know what you guys think. We can always revert or change if things change on the prettier end. Mostly wanted to verify this would actually work

**Comment by mgrip** [comment]
cc @evilebottnawi @czosel

**Comment by mgrip** [comment]
tests are passing! 🎉

**Comment by czosel** [comment]
@mgrip looks like deleting the travis cache did the job! :smile: 
Unfortunately, I don't have time to take a closer look at this right now - but it certainly looks great on first glance and I'm truly excited for our first release :wink:

**Comment by evilebottnawi** [comment]
@czosel we can merge this master and when fix after report users, we can handle right now all possible comments location :+1:

File: tests/case/case.php [hunk 4]
```diff
@@ -1,236 +1,4 @@
...
-
-ABSTRACT CLASS AbstractClass
-{
-    ABSTRACT PROTECTED FUNCTION getValue();
-    ABSTRACT PROTECTED FUNCTION prefixValue($prefix);
-
-    PUBLIC function printOut() {
-        PRINT $this->getValue();
-    }
+FOR ($i = 0; $i < 10; $i++) {
```
Comment (evilebottnawi) [inline tests/case/case.php]: It seems I accidentally deleted tests, need return /cc @mgrip

Comment (mgrip) [reply tests/case/case.php]: looks like there are a couple types of comments not getting printed when i add it back. ill get it updated and fixed, then I think we'll be good to merge 👍

**Comment by mgrip** [comment]
@evilebottnawi added back `case` test, updated 👍

**Comment by evilebottnawi** [comment]
@mgrip great work! Ready for review and merge?

**Comment by mgrip** [comment]
I think so! 🎉

File: src/printer.js [hunk 5]
```diff
@@ -1484,12 +1534,12 @@ function printNode(path, options, print) {
       }
       return concat(parts);
     }
-    case "doc": {
+    /*case "doc": {
```
Comment (evilebottnawi) [inline src/printer.js]: Let's remove this node :+1:

File: tests/comments/__snapshots__/jsfmt.spec.js.snap [hunk 6]
```diff
@@ -106,17 +106,100 @@ class Foo
         // there should be no docblock here
         $this->title = $title;
     }
+
+    /**
```
Comment (evilebottnawi) [inline tests/comments/__snapshots__/jsfmt.spec.js.snap]: Let's check `isDoc` comment (maybe use regex or package), if yes - pretty output, if no - print as is :+1:

**Comment by mgrip** [comment]
@evilebottnawi all set!

**Comment by evilebottnawi** [comment]
@mgrip approve :+1: let's wait what says @czosel and feel free to merge. Tomorrow i will test on big code base with comments and send PR with fixes :+1:

File: src/comments.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+const handleOwnLineComment = function(comment) {
+  if (
+    handleClass(comment) ||
+    handleFunction(comment) ||
+    handleForLoop(comment) ||
+    handleTryCatch(comment)
+  ) {
+    return true;
+  }
+  return false;
```
Comment (czosel) [inline src/comments.js]: Why not just `return handleClass(comment)  || ...` (no extra `if`)?

Comment (mgrip) [reply src/comments.js]: good point 😄 i'll update now

Comment (mgrip) [reply src/comments.js]: all set 👍

File: src/printer.js (hunk 2 above)
Comment (czosel) [reply src/printer.js]: Isn't this fixed in prettier core now?

Comment (czosel) [reply src/printer.js]: Apart from this, this looks all reaaally good! 💯

**Comment by mgrip** [comment]
I copied the `printDanglingComments` function from prettier since technically it is part of the printer, I think it makes sense to be plugin-specific. Think we should be all set!

**Comment by mgrip** [comment]
thanks everyone, this was a big one! 🎉

**Pull request merged**